bundles/org.openhab.voice.actiontemplatehli/src/main/java/org/openhab/voice/actiontemplatehli/internal/ActionTemplateInterpreter.java

```diff
@@ -0,0 +1,1159 @@
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.openhab.core.types.TypeParser;
+import org.openhab.core.types.UnDefType;
+import org.openhab.core.voice.text.HumanLanguageInterpreter;
+import org.openhab.core.voice.text.InterpretationException;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import opennlp.tools.dictionary.Dictionary;
+import opennlp.tools.lemmatizer.DictionaryLemmatizer;
+import opennlp.tools.lemmatizer.Lemmatizer;
+import opennlp.tools.lemmatizer.LemmatizerME;
+import opennlp.tools.lemmatizer.LemmatizerModel;
+import opennlp.tools.namefind.DictionaryNameFinder;
+import opennlp.tools.namefind.NameFinderME;
+import opennlp.tools.namefind.TokenNameFinderModel;
+import opennlp.tools.postag.POSDictionary;
+import opennlp.tools.postag.POSModel;
+import opennlp.tools.postag.POSTaggerME;
+import opennlp.tools.tokenize.SimpleTokenizer;
+import opennlp.tools.tokenize.Tokenizer;
+import opennlp.tools.tokenize.TokenizerME;
+import opennlp.tools.tokenize.TokenizerModel;
+import opennlp.tools.tokenize.WhitespaceTokenizer;
+import opennlp.tools.util.Span;
+import opennlp.tools.util.StringList;
+
+/**
+ * The {@link ActionTemplateInterpreter} is a configurable interpreter powered by OpenNLP
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME, description_uri = SERVICE_CATEGORY + ":"
+        + SERVICE_ID)
+public class ActionTemplateInterpreter implements HumanLanguageInterpreter {
+    static {
+        Logger logger = LoggerFactory.getLogger(ActionTemplateInterpreter.class);
+        createFolder(logger, NLP_FOLDER);
+        createFolder(logger, NER_FOLDER);
+        createFolder(logger, POS_FOLDER);
+        createFolder(logger, TYPE_ACTION_CONFIGS_FOLDER);
+    }
+    protected ActionTemplateInterpreterConfiguration config = new ActionTemplateInterpreterConfiguration();
+    private static final Pattern COLOR_HEX_PATTERN = Pattern.compile("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$");
+    private final Logger logger = LoggerFactory.getLogger(ActionTemplateInterpreter.class);
+    private final LocaleService localeService;
+    private final ItemRegistry itemRegistry;
+    private final MetadataRegistry metadataRegistry;
+    private final EventPublisher eventPublisher;
+    private Tokenizer tokenizer = WhitespaceTokenizer.INSTANCE;
+    private ExpiringCacheMap<String, NLPInfo> cacheMap = new ExpiringCacheMap<>(Duration.of(1, ChronoUnit.MINUTES));
```

Comment:
The cache is to avoid tokenizing again the same text. But why a one minute validity ?

Reply:
Yes sorry, it has an explanation.

My initial idea (but I decided not to do it for now) was to add some rule actions to this add-on to expose the getNLPInfo the NER and the POS to the rules so the user has like a little nlp framework to use on the rules. 

That is the reason the NLPInfo class and the related method was public (changed to private right now). The cache was intended to reuse the processing in case a rule calls getNLPInfo with the same text to do some custom processing there with the tokens or lemmas.

Maybe it has value on its own if we allow to customize the expiration time. WDYT?

Reply:
I rather understand this cache is now useless. 
So I would suggest to remove it.